# Load data

In [ ]:
# load training data
import gdown
import pandas as pd

file_id = '15fxpr_i_w47WFLxq-UQn9LL0kmvETMeB'
gdown.download(f'https://drive.google.com/uc?id={file_id}', 'insurance_train.csv', quiet=False)

df_train = pd.read_csv('insurance_train.csv')
pd.set_option('display.max_columns', None)
df_train

Downloading...
From: https://drive.google.com/uc?id=15fxpr_i_w47WFLxq-UQn9LL0kmvETMeB
To: /content/insurance_train.csv
100%|██████████| 4.90M/4.90M [00:00<00:00, 28.2MB/s]


,X.1,X.2,X.3,X.4,X.5,X.6,X.7,X.8,X.9,X.10,X.11,X.12,X.13,X.14,X.15,X.16,X.17,X.18,X.19,X.20,X.21,X.22,X.23,X.24,X.25,X.26,X.27,X.28
0,34429,01/06/2017,01/06/2017,01/06/2018,29/08/1996,31/05/2016,0,1,1,1,1,1,0,573.07,0.0,0,0,0.00,3,0,0,1994,75,1360,12795.55,4,P,1040
1,5552,19/09/2016,19/09/2018,19/09/2019,30/04/1992,03/08/2010,0,3,1,1,1,0,0,299.19,0.0,0,0,0.00,3,0,0,2004,100,1896,15386.00,5,D,1165
2,47700,08/01/2003,08/01/2018,08/01/2019,23/03/1972,01/02/1998,1,20,2,2,1,0,0,257.66,0.0,0,2,0.13,3,0,0,2002,95,1390,16470.00,5,P,1235
3,25425,01/10/2015,01/10/2018,01/10/2019,05/03/1946,07/08/1964,0,8,2,3,1,0,0,288.70,0.0,0,1,0.00,3,0,0,2010,140,2497,27381.00,5,D,1984
4,4727,26/01/2017,26/01/2018,26/01/2019,25/04/1973,24/07/1998,0,9,1,1,1,0,0,741.16,0.0,0,3,1.08,3,0,0,1998,80,1969,36600.00,3,P,1702
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37446,19562,26/06/2018,26/06/2018,26/06/2019,02/10/1960,02/11/1980,1,1,1,1,1,0,0,402.66,0.0,0,0,0.00,2,0,0,2008,84,1896,23977.20,5,D,1755
37447,5888,23/05/2001,23/05/2018,23/05/2019,01/10/1944,01/04/1963,0,18,1,1,1,0,0,84.43,0.0,0,0,0.00,4,0,0,2001,81,4525,42920.00,0,D,4525
37448,44775,01/02/2016,01/02/2018,01/02/2019,16/04/1992,05/05/2011,0,3,1,1,1,0,0,254.92,0.0,0,0,0.00,3,1,0,1996,55,998,8426.19,3,P,785
37449,35358,24/05/2016,24/05/2017,24/05/2018,20/12/1975,14/10/1996,0,2,1,1,1,1,0,89.38,0.0,0,0,0.00,1,0,0,2005,9,125,2165.00,0,P,105


In [ ]:

# Test set file id
file_id_test = '1EriQzt5qvAG3FB8eWXYk8K96ed12UvPQ'  # <-- this is the insurance_test.csv file id

# Download and read test set
gdown.download(f'https://drive.google.com/uc?id={file_id_test}', 'insurance_test.csv', quiet=False)

df_test = pd.read_csv('insurance_test.csv')
pd.set_option('display.max_columns', None)

df_test.columns


Downloading...
From: https://drive.google.com/uc?id=1EriQzt5qvAG3FB8eWXYk8K96ed12UvPQ
To: /content/insurance_test.csv
100%|██████████| 1.77M/1.77M [00:00<00:00, 160MB/s]


Index(['X.2', 'X.3', 'X.4', 'X.5', 'X.6', 'X.7', 'X.8', 'X.9', 'X.10', 'X.11',
       'X.12', 'X.13', 'X.14', 'X.19', 'X.20', 'X.21', 'X.22', 'X.23', 'X.24',
       'X.25', 'X.26', 'X.27', 'X.28'],
      dtype='object')

In [ ]:
# check for missing val
df_train.isnull().sum()


,0
X.1,0
X.2,0
X.3,0
X.4,0
X.5,0
X.6,0
X.7,0
X.8,0
X.9,0
X.10,0


# Data Cleaning

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Define the y cols
df_train['CS'] = (df_train['X.16'] > 0).astype(int)  #ty x15 or 16
df_train['LC'] = 0.0
df_train['HALC'] = 0.0
nonzero_mask = df_train['X.16'] > 0
df_train.loc[nonzero_mask, 'LC'] = df_train.loc[nonzero_mask, 'X.15'] / df_train.loc[nonzero_mask, 'X.16']
df_train.loc[nonzero_mask, 'HALC'] = df_train.loc[nonzero_mask, 'LC'] * df_train.loc[nonzero_mask, 'X.18']


#Dropping Columns
cols_to_drop = ['X.1', 'X.15', 'X.16', 'X.17', 'X.18']
df_train.drop(columns=[col for col in cols_to_drop if col in df_train.columns], inplace=True)

# Convert date columns to datetime
date_cols = ['X.2', 'X.3', 'X.4', 'X.5', 'X.6']
for col in date_cols:
    df_train[col] = pd.to_datetime(df_train[col], errors='coerce', dayfirst=True)


# Feature engineering
df_train['vehicle_age'] = 2020 - df_train['X.22']                    # from registration year
df_train['age'] = 2020 - df_train['X.5'].dt.year                     # from birthdate
df_train['driving_experience'] = 2020 - df_train['X.6'].dt.year     # from license issue date
df_train['policy_duration_days'] = (df_train['X.4'] - df_train['X.2']).dt.days
df_train['time_since_last_renewal_days'] = (df_train['X.3'] - df_train['X.2']).dt.days

# Drop raw date columns from both train and test
df_train.drop(columns=[col for col in date_cols if col in df_train.columns], inplace=True)

# One-hot encoding
categorical_cols = ['X.7', 'X.13', 'X.19', 'X.20', 'X.21', 'X.27']
df_train = pd.get_dummies(df_train, columns=categorical_cols, drop_first=True)





# Task 1

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import TweedieRegressor
from sklearn.metrics import mean_squared_error

df=df_train
# Drop raw dates and leakage columns
drop_cols = ["LC", "HALC", "CS"]

# Filter out extreme outliers
q_lc = df["LC"].quantile([0.001, 0.999])
q_halc = df["HALC"].quantile([0.001, 0.999])
#q_lc = df["LC"].quantile([0.01, 0.99])
#q_halc = df["HALC"].quantile([0.01, 0.99])
df_filtered = df[(df["LC"] >= q_lc.iloc[0]) & (df["LC"] <= q_lc.iloc[1]) &
               (df["HALC"] >= q_halc.iloc[0]) & (df["HALC"] <= q_halc.iloc[1])]

# Prepare features and targets
X = df_filtered.drop(columns=drop_cols)
#X = pd.get_dummies(X, columns=["X.27"], drop_first=True)
y_lc = df_filtered["LC"]
y_halc = df_filtered["HALC"]

# Split and scale
X_train_lc, X_val_lc, y_train_lc, y_val_lc = train_test_split(X, y_lc, test_size=0.2, random_state=42)
X_train_halc, X_val_halc, y_train_halc, y_val_halc = train_test_split(X, y_halc, test_size=0.2, random_state=42)

scaler_lc = StandardScaler().fit(X_train_lc)
X_train_lc_scaled = scaler_lc.transform(X_train_lc)
X_val_lc_scaled = scaler_lc.transform(X_val_lc)

scaler_halc = StandardScaler().fit(X_train_halc)
X_train_halc_scaled = scaler_halc.transform(X_train_halc)
X_val_halc_scaled = scaler_halc.transform(X_val_halc)

# Fit Tweedie Regressors
glm_lc = TweedieRegressor(power=1.5, link='log', alpha=0.0, max_iter=1000)
glm_lc.fit(X_train_lc_scaled, y_train_lc)
y_pred_lc = glm_lc.predict(X_val_lc_scaled)
mse_lc = mean_squared_error(y_val_lc, y_pred_lc)
rmse_lc = np.sqrt(mse_lc)

glm_halc = TweedieRegressor(power=1.5, link='log', alpha=0.0, max_iter=1000)
glm_halc.fit(X_train_halc_scaled, y_train_halc)
y_pred_halc = glm_halc.predict(X_val_halc_scaled)
mse_halc = mean_squared_error(y_val_halc, y_pred_halc)
rmse_halc = np.sqrt(mse_halc)

print(f"LC RMSE: {rmse_lc:.2f}")
print(f"HALC RMSE: {rmse_halc:.2f}")

/usr/local/lib/python3.12/dist-packages/numpy/_core/_methods.py:52: RuntimeWarning: invalid value encountered in reduce
  return umr_sum(a, axis, dtype, out, keepdims, initial, where)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_linear_loss.py:330: RuntimeWarning: invalid value encountered in matmul
  grad[:n_features] = X.T @ grad_pointwise + l2_reg_strength * weights
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_linear_loss.py:330: RuntimeWarning: invalid value encountered in matmul
  grad[:n_features] = X.T @ grad_pointwise + l2_reg_strength * weights
/usr/local/lib/python3.12/dist-packages/numpy/_core/_methods.py:52: RuntimeWarning: invalid value encountered in reduce
  return umr_sum(a, axis, dtype, out, keepdims, initial, where)


LC RMSE: 291.00
HALC RMSE: 573.65


In [ ]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train_lc, y_train_lc)

y_pred_rf = rf.predict(X_val_lc)
mse_rf = mean_squared_error(y_val_lc, y_pred_rf)
rmse_rf = np.sqrt(mse_rf)

print("Random Forest - LC")
print(f"MSE: {mse_rf:.2f}")
print(f"RMSE: {rmse_rf:.2f}")

# Reuse same features
rf_halc = RandomForestRegressor(n_estimators=100, random_state=42)
rf_halc.fit(X_train_halc, y_train_halc)

y_pred_rf_halc = rf_halc.predict(X_val_halc)
mse_rf_halc = mean_squared_error(y_val_halc, y_pred_rf_halc)
rmse_rf_halc = np.sqrt(mse_rf_halc)

print("Random Forest - HALC")
print(f"MSE: {mse_rf_halc:.2f}")
print(f"RMSE: {rmse_rf_halc:.2f}")

Random Forest - LC
MSE: 88761.47
RMSE: 297.93
Random Forest - HALC
MSE: 339787.00
RMSE: 582.91


In [ ]:
from sklearn.ensemble import GradientBoostingRegressor

gb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, random_state=42)
gb.fit(X_train_lc, y_train_lc)

y_pred_gb = gb.predict(X_val_lc)
mse_gb = mean_squared_error(y_val_lc, y_pred_gb)
rmse_gb = np.sqrt(mse_gb)

print("Gradient Boosting - LC")
print(f"MSE: {mse_gb:.2f}")
print(f"RMSE: {rmse_gb:.2f}")

gb_halc = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, random_state=42)
gb_halc.fit(X_train_halc, y_train_halc)

y_pred_gb_halc = gb_halc.predict(X_val_halc)
mse_gb_halc = mean_squared_error(y_val_halc, y_pred_gb_halc)
rmse_gb_halc = np.sqrt(mse_gb_halc)

print("Gradient Boosting - HALC")
print(f"MSE: {mse_gb_halc:.2f}")
print(f"RMSE: {rmse_gb_halc:.2f}")

Gradient Boosting - LC
MSE: 83610.06
RMSE: 289.15
Gradient Boosting - HALC
MSE: 324290.07
RMSE: 569.46


In [ ]:
# @title
from lightgbm import LGBMRegressor

lgb = LGBMRegressor(n_estimators=100, learning_rate=0.1, random_state=42)
lgb.fit(X_train_lc, y_train_lc)

y_pred_lgb = lgb.predict(X_val_lc)
mse_lgb = mean_squared_error(y_val_lc, y_pred_lgb)
rmse_lgb = np.sqrt(mse_lgb)

print("LightGBM - LC")
print(f"MSE: {mse_lgb:.2f}")
print(f"RMSE: {rmse_lgb:.2f}")

lgb_halc = LGBMRegressor(n_estimators=100, tweedie_variance_power= 1.5, learning_rate=0.1, random_state=42)
lgb_halc.fit(X_train_halc, y_train_halc)

y_pred_lgb_halc = lgb_halc.predict(X_val_halc)
mse_lgb_halc = mean_squared_error(y_val_halc, y_pred_lgb_halc)
rmse_lgb_halc = np.sqrt(mse_lgb_halc)

print("LightGBM - HALC")
print(f"MSE: {mse_lgb_halc:.2f}")
print(f"RMSE: {rmse_lgb_halc:.2f}")

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.006010 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1935
[LightGBM] [Info] Number of data points in the train set: 29921, number of used features: 25
[LightGBM] [Info] Start training from score 51.995538
LightGBM - LC
MSE: 82999.90
RMSE: 288.10
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.005572 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1935
[LightGBM] [Info] Number of data points in the train set: 29921, number of used features: 25
[LightGBM] [Info] Start training from score 92.570930
LightGBM - HALC
MSE: 323034.69
RMSE: 568.36


In [ ]:
from xgboost import XGBRegressor

xgb = XGBRegressor(n_estimators=100, learning_rate=0.1, random_state=42, verbosity=0)
xgb.fit(X_train_lc, y_train_lc)

y_pred_xgb = xgb.predict(X_val_lc)
mse_xgb = mean_squared_error(y_val_lc, y_pred_xgb)
rmse_xgb = np.sqrt(mse_xgb)

print("XGBoost - LC")
print(f"MSE: {mse_xgb:.2f}")
print(f"RMSE: {rmse_xgb:.2f}")

xgb_halc = XGBRegressor(n_estimators=100, learning_rate=0.1, random_state=42, verbosity=0)
xgb_halc.fit(X_train_halc, y_train_halc)

y_pred_xgb_halc = xgb_halc.predict(X_val_halc)
mse_xgb_halc = mean_squared_error(y_val_halc, y_pred_xgb_halc)
rmse_xgb_halc = np.sqrt(mse_xgb_halc)

print("XGBoost - HALC")
print(f"MSE: {mse_xgb_halc:.2f}")
print(f"RMSE: {rmse_xgb_halc:.2f}")

XGBoost - LC
MSE: 83034.96
RMSE: 288.16
XGBoost - HALC
MSE: 324819.54
RMSE: 569.93


In [ ]:
from sklearn.neural_network import MLPRegressor

# Scale features again (in case you’ve rerun anything)
scaler_nn = StandardScaler()
X_train_scaled_lc = scaler_nn.fit_transform(X_train_lc)
X_val_scaled_lc = scaler_nn.transform(X_val_lc)

# Fit neural network
nn = MLPRegressor(hidden_layer_sizes=(64, 32), max_iter=300, random_state=42)
nn.fit(X_train_scaled_lc, y_train_lc)

y_pred_nn = nn.predict(X_val_scaled_lc)
mse_nn = mean_squared_error(y_val_lc, y_pred_nn)
rmse_nn = np.sqrt(mse_nn)

print("Neural Net - LC")
print(f"MSE: {mse_nn:.2f}")
print(f"RMSE: {rmse_nn:.2f}")

# Scale features again (in case you’ve rerun anything)
scaler_nn = StandardScaler()
X_train_scaled_halc = scaler_nn.fit_transform(X_train_halc)
X_val_scaled_halc = scaler_nn.transform(X_val_halc)


nn_halc = MLPRegressor(hidden_layer_sizes=(64, 32), max_iter=300, random_state=42)
nn_halc.fit(X_train_scaled_halc, y_train_halc)

y_pred_nn_halc = nn_halc.predict(X_val_scaled_halc)
mse_nn_halc = mean_squared_error(y_val_halc, y_pred_nn_halc)
rmse_nn_halc = np.sqrt(mse_nn_halc)

print("Neural Net - HALC")
print(f"MSE: {mse_nn_halc:.2f}")
print(f"RMSE: {rmse_nn_halc:.2f}")

/usr/local/lib/python3.12/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Net - LC
MSE: 85931.46
RMSE: 293.14
Neural Net - HALC
MSE: 324680.99
RMSE: 569.81


/usr/local/lib/python3.12/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [ ]:
# Create a summary DataFrame with your actual variable names
results_summary = pd.DataFrame({
    'Model': ['GLM (Tweedie)', 'Random Forest', 'Gradient Boosting', 'LightGBM', 'XGBoost', 'Neural Network'],
    'LC MSE': [mse_lc, mse_rf, mse_gb, mse_lgb, mse_xgb, mse_nn],
    'LC RMSE': [rmse_lc, rmse_rf, rmse_gb, rmse_lgb, rmse_xgb, rmse_nn],
    'HALC MSE': [mse_halc, mse_rf_halc, mse_gb_halc, mse_lgb_halc, mse_xgb_halc, mse_nn_halc],
    'HALC RMSE': [rmse_halc, rmse_rf_halc, rmse_gb_halc, rmse_lgb_halc, rmse_xgb_halc, rmse_nn_halc]
})

# Round values for readability
results_summary = results_summary.round(2)

# Highlight the best RMSEs
def highlight_best(s):
    is_min = s == s.min()
    return ['background-color: lightgreen' if v else '' for v in is_min]

# Apply styling
styled_results = results_summary.style.apply(highlight_best, subset=['LC RMSE', 'HALC RMSE'])
styled_results

,Model,LC MSE,LC RMSE,HALC MSE,HALC RMSE
0,GLM (Tweedie),84680.950000,291.000000,329071.410000,573.650000
1,Random Forest,88761.470000,297.930000,339787.000000,582.910000
2,Gradient Boosting,83610.060000,289.150000,324290.070000,569.460000
3,LightGBM,82999.900000,288.100000,323034.690000,568.360000
4,XGBoost,83034.960000,288.160000,324819.540000,569.930000
5,Neural Network,85931.460000,293.140000,324680.990000,569.810000


In [ ]:
from sklearn.model_selection import KFold
from sklearn.linear_model import TweedieRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.neural_network import MLPRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
import numpy as np
import pandas as pd

kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Models
models = {
    "GLM (Tweedie)": TweedieRegressor(power=1.5, link='log', alpha=0.0, max_iter=1000),
    "Random Forest": RandomForestRegressor(n_estimators=100, random_state=42),
    "Gradient Boosting": GradientBoostingRegressor(n_estimators=100, random_state=42),
    "LightGBM": LGBMRegressor(n_estimators=100, random_state=42),
    "XGBoost": XGBRegressor(n_estimators=100, random_state=42, verbosity=0),
    "Neural Network": MLPRegressor(hidden_layer_sizes=(64, 32), max_iter=300, random_state=42)
}

# Store results
cv_results = []

# Loop through models
for name, model in models.items():
    print(f"Running CV for: {name}")

    # Scale only for GLM and MLP
    if name in ['GLM (Tweedie)', 'Neural Network']:
        scaler = StandardScaler()
        X_scaled = scaler.fit_transform(X)
    else:
        X_scaled = X.values  # no scaling for tree models

    # LC
    mse_lc_list, rmse_lc_list = [], []
    for train_idx, val_idx in kf.split(X_scaled):
        X_train, X_val = X_scaled[train_idx], X_scaled[val_idx]
        y_train, y_val = y_lc.iloc[train_idx], y_lc.iloc[val_idx]

        model.fit(X_train, y_train)
        preds = model.predict(X_val)

        mse = mean_squared_error(y_val, preds)
        rmse = np.sqrt(mse)
        mse_lc_list.append(mse)
        rmse_lc_list.append(rmse)

    # HALC
    mse_halc_list, rmse_halc_list = [], []
    for train_idx, val_idx in kf.split(X_scaled):
        X_train, X_val = X_scaled[train_idx], X_scaled[val_idx]
        y_train, y_val = y_halc.iloc[train_idx], y_halc.iloc[val_idx]

        model.fit(X_train, y_train)
        preds = model.predict(X_val)

        mse = mean_squared_error(y_val, preds)
        rmse = np.sqrt(mse)
        mse_halc_list.append(mse)
        rmse_halc_list.append(rmse)

    # Store results
    cv_results.append({
        "Model": name,
        "LC MSE": np.mean(mse_lc_list),
        "LC RMSE": np.mean(rmse_lc_list),
        "HALC MSE": np.mean(mse_halc_list),
        "HALC RMSE": np.mean(rmse_halc_list)
    })

# Create summary DataFrame
cv_summary = pd.DataFrame(cv_results).round(2)
cv_summary = cv_summary.sort_values(by="HALC RMSE")

# Highlight best RMSEs
def highlight_best(s):
    is_min = s == s.min()
    return ['background-color: lightgreen' if v else '' for v in is_min]

styled_cv_summary = cv_summary.style.apply(highlight_best, subset=['LC RMSE', 'HALC RMSE'])

styled_cv_summary

Running CV for: GLM (Tweedie)


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_linear_loss.py:330: RuntimeWarning: invalid value encountered in matmul
  grad[:n_features] = X.T @ grad_pointwise + l2_reg_strength * weights
/usr/local/lib/python3.12/dist-packages/numpy/_core/_methods.py:52: RuntimeWarning: invalid value encountered in reduce
  return umr_sum(a, axis, dtype, out, keepdims, initial, where)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_linear_loss.py:330: RuntimeWarning: invalid value encountered in matmul
  grad[:n_features] = X.T @ grad_pointwise + l2_reg_strength * weights
/usr/local/lib/python3.12/dist-packages/numpy/_core/_methods.py:52: RuntimeWarning: invalid value encountered in reduce
  return umr_sum(a, axis, dtype, out, keepdims, initial, where)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_linear_loss.py:330: RuntimeWarning: invalid value encountered in matmul
  grad[:n_features] = X.T @ grad_pointwise + l2_reg_strength * weights
/usr/local/lib/py

Running CV for: Random Forest


In [ ]:
# prompt: do not showing warnings

import warnings
warnings.filterwarnings('ignore')


LightGBM Hyperparameter Tuning (HALC Clipped)

In [ ]:
from lightgbm import LGBMRegressor
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error
import numpy as np

# Input data
X_values = X.values

# Cross-validation setup
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Parameter grid
param_grid = {
    'n_estimators': [100, 300],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.05, 0.1]
}

# To store results
tuning_results = []

# Grid search
for n in param_grid['n_estimators']:
    for d in param_grid['max_depth']:
        for lr in param_grid['learning_rate']:
            model = LGBMRegressor(n_estimators=n, max_depth=d, learning_rate=lr, random_state=42)

            mse_list, rmse_list = [], []

            for train_idx, val_idx in kf.split(X_values):
                X_train, X_val = X_values[train_idx], X_values[val_idx]
                y_train, y_val = y_lc.iloc[train_idx], y_lc.iloc[val_idx]

                model.fit(X_train, y_train)
                preds = model.predict(X_val)

                mse = mean_squared_error(y_val, preds)
                rmse = np.sqrt(mse)
                mse_list.append(mse)
                rmse_list.append(rmse)

            tuning_results.append({
                'n_estimators': n,
                'max_depth': d,
                'learning_rate': lr,
                'Mean MSE': np.mean(mse_list),
                'Mean RMSE': np.mean(rmse_list)
            })

# Results table
tuning_df = pd.DataFrame(tuning_results).sort_values(by='Mean RMSE').round(2)
tuning_df.reset_index(drop=True, inplace=True)
tuning_df

##Cross Validation

In [ ]:
from lightgbm import LGBMRegressor
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error
import numpy as np

# Input data
X_values = X.values

# Cross-validation setup
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Parameter grid
param_grid = {
    'n_estimators': [100, 300],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.05, 0.1]
}

# To store results
tuning_results = []

# Grid search
for n in param_grid['n_estimators']:
    for d in param_grid['max_depth']:
        for lr in param_grid['learning_rate']:
            model = LGBMRegressor(n_estimators=n, max_depth=d, learning_rate=lr, random_state=42)

            mse_list, rmse_list = [], []

            for train_idx, val_idx in kf.split(X_values):
                X_train, X_val = X_values[train_idx], X_values[val_idx]
                y_train, y_val = y_halc.iloc[train_idx], y_halc.iloc[val_idx]

                model.fit(X_train, y_train)
                preds = model.predict(X_val)

                mse = mean_squared_error(y_val, preds)
                rmse = np.sqrt(mse)
                mse_list.append(mse)
                rmse_list.append(rmse)

            tuning_results.append({
                'n_estimators': n,
                'max_depth': d,
                'learning_rate': lr,
                'Mean MSE': np.mean(mse_list),
                'Mean RMSE': np.mean(rmse_list)
            })

# Results table
tuning_df = pd.DataFrame(tuning_results).sort_values(by='Mean RMSE').round(2)
tuning_df.reset_index(drop=True, inplace=True)
tuning_df

We performed hyperparameter tuning using grid search across LightGBM’s key parameters, and found that a combination of 100 estimators, max depth of 7, and a learning rate of 0.05 produced the lowest RMSE (633.64) for clipped HALC, improving predictive performance over default settings.

##SHAP interpretability

In [ ]:
final_lgbm = LGBMRegressor(
    n_estimators=100,
    max_depth=7,
    learning_rate=0.05,
    random_state=42
)

final_lgbm.fit(X, y_halc)

#SHAP for HALC with LightGBM
import shap

# Initialize SHAP explainer for LightGBM
explainer_halc = shap.Explainer(final_lgbm)

# Compute SHAP values
shap_values_halc = explainer_halc(X)

# Plot summary
shap.summary_plot(shap_values_halc, X, plot_type="bar")

# Detailed beeswarm SHAP plot
shap.summary_plot(shap_values_halc, X)

In [ ]:
from lightgbm import LGBMRegressor

lgbm_lc = LGBMRegressor(n_estimators=100, max_depth=5, learning_rate=0.1, random_state=42)
lgbm_lc.fit(X, y_lc)

In [ ]:
import shap

# Create SHAP explainer
explainer_lc = shap.Explainer(lgbm_lc)

# Compute SHAP values
shap_values_lc = explainer_lc(X)

# Bar plot
shap.summary_plot(shap_values_lc, X, plot_type="bar")

# Detailed summary plot
shap.summary_plot(shap_values_lc, X)

##Log Transformation

In [ ]:
y_lc_log = np.log1p(df_filtered["LC"])
y_halc_log = np.log1p(df_filtered["HALC"])

# Step 2: Train/test split
X_train_lc_log, X_val_lc_log, y_train_lc_log, y_val_lc_log = train_test_split(X, y_lc_log, test_size=0.2, random_state=42)
X_train_halc_log, X_val_halc_log, y_train_halc_log, y_val_halc_log = train_test_split(X, y_halc_log, test_size=0.2, random_state=42)

# Step 3: Train models
model_lc_log = LGBMRegressor(n_estimators=100, learning_rate=0.1, random_state=42)
model_lc_log.fit(X_train_lc_log, y_train_lc_log)

model_halc_log = LGBMRegressor(n_estimators=100, learning_rate=0.1, tweedie_variance_power=1.5, random_state=42)
model_halc_log.fit(X_train_halc_log, y_train_halc_log)

# Step 4: Predict and inverse-transform
y_pred_lc_log = model_lc_log.predict(X_val_lc_log)
y_pred_lc_with_log = np.expm1(y_pred_lc_log)
y_val_lc_after_log = np.expm1(y_val_lc_log)

y_pred_halc_log = model_halc_log.predict(X_val_halc_log)
y_pred_halc_with_log = np.expm1(y_pred_halc_log)
y_val_halc_after_log = np.expm1(y_val_halc_log)

# Step 5: Evaluate
rmse_lc_log = np.sqrt(mean_squared_error(y_val_lc_after_log, y_pred_lc_with_log))
rmse_halc_log = np.sqrt(mean_squared_error(y_val_halc_after_log, y_pred_halc_with_log))

print(f"✅ LC RMSE (log-transformed): {rmse_lc_log:.2f}")
print(f"✅ HALC RMSE (log-transformed): {rmse_halc_log:.2f}")

Using Log Transformation did not improve RMSE

##Selected Features

In [ ]:
df2=df_train
# Drop raw dates and leakage columns
drop_cols = ["LC", "HALC", "CS", 'X.9', 'X.7_1', 'X.20_1', 'X.27_P', 'X.26', 'X.21_1']

# Filter out extreme outliers
q_lc_selected = df2["LC"].quantile([0.001, 0.999])
q_halc_selected = df2["HALC"].quantile([0.001, 0.999])
df_filtered_selected = df2[(df2["LC"] >= q_lc_selected.iloc[0]) & (df2["LC"] <= q_lc_selected.iloc[1]) &
               (df2["HALC"] >= q_halc_selected.iloc[0]) & (df2["HALC"] <= q_halc_selected.iloc[1])]

# Prepare features and targets
X_selected = df_filtered_selected.drop(columns=drop_cols)
#X = pd.get_dummies(X, columns=["X.27"], drop_first=True)
y_lc_selected = df_filtered_selected["LC"]
y_halc_selected = df_filtered_selected["HALC"]


# Step 4: Train/test split
X_train_lc_selected, X_test_lc_selected, y_train_lc_selected, y_test_lc_selected = train_test_split(X_selected, y_lc_selected, test_size=0.2, random_state=42)
X_train_halc_selected, X_test_halc_selected, y_train_halc_selected, y_test_halc_selected = train_test_split(X_selected, y_halc_selected, test_size=0.2, random_state=42)

# Step 5: Train LightGBM
model_lc_selected = LGBMRegressor(n_estimators=100, learning_rate=0.1, random_state=42)
model_lc_selected.fit(X_train_lc_selected, y_train_lc_selected)

model_halc_selected = LGBMRegressor(n_estimators=100, learning_rate=0.1, tweedie_variance_power=1.5, random_state=42)
model_halc_selected.fit(X_train_halc_selected, y_train_halc_selected)

# Step 6: Predict and evaluate
y_pred_lc_selected = model_lc_selected.predict(X_test_lc_selected)
y_pred_halc_selected = model_halc_selected.predict(X_test_halc_selected)

rmse_lc_selected = np.sqrt(mean_squared_error(y_test_lc_selected, y_pred_lc_selected))
rmse_halc_selected = np.sqrt(mean_squared_error(y_test_halc_selected, y_pred_halc_selected))

print(f"✅ LC RMSE: {rmse_lc_selected:.2f}")
print(f"✅ HALC RMSE: {rmse_halc_selected:.2f}")


Without selected features -

LC RMSE: 288.10

HALC RMSE: 568.36

With selected features did not improve RMSE, so we wil not use it

##Analysis

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# X_val is already a numpy array, no need to access .values
y_pred_halc = final_lgbm.predict(X_val)

# Ensure y_val_halc and y_pred_halc have the same length
y_val_halc = y_val_halc.iloc[:len(y_pred_halc)]  # Adjust the length of y_val_halc to match y_pred_halc

# Now create the scatterplot
sns.scatterplot(x=y_val_halc, y=y_pred_halc, alpha=0.5)
plt.xlabel("Actual HALC")
plt.ylabel("Predicted HALC")
plt.title("Predicted vs Actual - HALC (LightGBM)")
plt.plot([y_val_halc.min(), y_val_halc.max()], [y_val_halc.min(), y_val_halc.max()], 'r--')
plt.grid(True)
plt.tight_layout()
plt.show()

In [ ]:
#Distribution of Prediction Errors HACL
errors_halc = y_val_halc - y_pred_halc

sns.histplot(errors_halc, bins=40, kde=True)
plt.title("Distribution of Prediction Errors - HALC")
plt.xlabel("Prediction Error (Actual - Predicted)")
plt.ylabel("Frequency")
plt.grid(True)
plt.tight_layout()
plt.show()

In [ ]:
# residual vs predicted HALC
residuals = y_val_halc - y_pred_halc

plt.scatter(y_pred_halc, residuals, alpha=0.4)
plt.axhline(0, color='red', linestyle='--')
plt.xlabel("Predicted HALC")
plt.ylabel("Residual (Actual - Predicted)")
plt.title("Residuals vs Predicted HALC")
plt.grid(True)
plt.tight_layout()
plt.show()

In [ ]:
# prompt: create a Predicted vs Actual - LC (LightGBM) graph LC

import matplotlib.pyplot as plt
import seaborn as sns

# Assuming y_val_lc and y_pred_lgb are already defined from your previous code
# Replace these with your actual variable names if different

plt.figure(figsize=(8, 6))
sns.scatterplot(x=y_val_lc, y=y_pred_lgb, alpha=0.5)
plt.xlabel("Actual LC")
plt.ylabel("Predicted LC")
plt.title("Predicted vs Actual - LC (LightGBM)")
plt.plot([y_val_lc.min(), y_val_lc.max()], [y_val_lc.min(), y_val_lc.max()], 'r--')  # Add a diagonal line
plt.grid(True)
plt.show()


In [ ]:
import matplotlib.pyplot as plt
#Distribution of Prediction Errors LC
errors_lc = y_val_lc - y_pred_lc

sns.histplot(errors_lc, bins=40, kde=True)
plt.title("Distribution of Prediction Errors - LC")
plt.xlabel("Prediction Error (Actual - Predicted)")
plt.ylabel("Frequency")
plt.grid(True)
plt.tight_layout()
plt.show()


In [ ]:
# prompt: create a Residuals vs Predicted LC graph

import matplotlib.pyplot as plt
# residual vs predicted LC
residuals_lc = y_val_lc - y_pred_lgb

plt.scatter(y_pred_lgb, residuals_lc, alpha=0.4)
plt.axhline(0, color='red', linestyle='--')
plt.xlabel("Predicted LC")
plt.ylabel("Residual (Actual - Predicted)")
plt.title("Residuals vs Predicted LC")
plt.grid(True)
plt.tight_layout()
plt.show()

In [ ]:
# Select example indices (replace with actual policyholder indices you want to display)
example_indices = [10, 25]

# Loop to generate Risk Profile Cards
for idx in example_indices:
    row = X_val_halc.iloc[[idx]]  # double brackets to keep as DataFrame
    true_val = y_val.iloc[idx]
    pred_val = final_lgbm.predict(row)[0]

    shap_values = explainer_halc(row)
    shap_df = pd.DataFrame({
        'Feature': row.columns,
        'SHAP Impact': shap_values.values[0],
        'Value': row.values[0]
    }).sort_values(by='SHAP Impact', key=abs, ascending=False).head(3)

    # Clean formatting
    print("\n" + "="*42)
    print(f"Risk Profile — Policyholder #{idx}")
    print("-"*42)
    print(f"Predicted HALC:  ${pred_val:,.2f}")
    print(f"Actual HALC:     ${true_val:,.2f}")
    print("-"*42)
    print("Top Feature Drivers:")
    print(shap_df[['Feature', 'Value', 'SHAP Impact']].to_string(index=False))
    print("="*42)

In [ ]:
# Predict on full validation set
y_pred_all = final_lgbm.predict(X_val)

# Set your tolerance for "accurate" prediction (e.g., ±$100)
tolerance = 100

# Classify
diff = y_pred_all - y_val_halc
overpredicted = (diff > tolerance).sum()
underpredicted = (diff < -tolerance).sum()
accurate = ((diff >= -tolerance) & (diff <= tolerance)).sum()

# Summary
print("Prediction Summary:")
print(f"✅ Accurate Predictions:    {accurate}")
print(f"🔼 Overpredicted:           {overpredicted}")
print(f"🔽 Underpredicted:          {underpredicted}")

In [ ]:
import matplotlib.pyplot as plt

labels = ['Accurate', 'Overpredicted', 'Underpredicted']
sizes = [accurate, overpredicted, underpredicted]
colors = ['lightgreen', 'salmon', 'lightblue']

plt.figure(figsize=(5,5))
plt.pie(sizes, labels=labels, autopct='%1.1f%%', colors=colors, startangle=140)
plt.title('Model Prediction Breakdown (Validation Set)')
plt.axis('equal')
plt.show()

In [ ]:
print(final_lgbm)
print(X_val.shape, y_val_halc.shape)
print("Tolerance used:", tolerance)

#Task 2

In [ ]:
# Define X and Y then split and standardize
X = df_train.drop(columns=['CS', 'HALC', 'LC'])

# Split into train and test set
X_train, X_test = train_test_split(X, test_size=0.2, random_state=42)

# y train cols
y_cs_train = df_train.loc[X_train.index,'CS']

# y test cols
y_cs_test = df_train.loc[X_test.index,'CS']

# Standardized X
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)


In [ ]:
# Check the distributuion 0/1
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

def check_class_balance(y, plot=True, threshold=0.4):
    y_series = pd.Series(y, name="label")
    counts = y_series.value_counts().sort_index()
    props  = counts / counts.sum()

    print("Class counts")
    print(counts, "\n")
    print("Class proportions")
    print(props.apply(lambda p: f"{p:.2%}"), "\n")

    minority_prop = props.min()
    if minority_prop < threshold or minority_prop > 1 - threshold:
        print(f"  Data appears imbalanced "
              f"(minority class = {minority_prop:.2%} < {threshold:.0%}).")
    else:
        print(" Data looks reasonably balanced.")

    if plot:
        sns.barplot(x=counts.index.astype(str), y=counts.values, palette="Set2")
        plt.ylabel("Number of samples")
        plt.xlabel("Class label")
        plt.title("Class distribution")
        plt.show()

    return counts, props , counts.sum()

counts, proportions, total_count = check_class_balance(y_cs_train, plot=True, threshold=0.4)


#Logistic Regression

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.metrics import roc_auc_score, recall_score, precision_score
from sklearn.linear_model import LogisticRegression

# Create logistic regression model
model = LogisticRegression(C=1, penalty='l1', max_iter=1000, solver='liblinear', class_weight='balanced')

# Fit model
model.fit(X_train, y_cs_train)

# Predict probabilities
y_proba_train = model.predict_proba(X_train)[:, 1]
y_proba_test_lr = model.predict_proba(X_test)[:, 1]

# AUC for train/test
auc_train = roc_auc_score(y_cs_train, y_proba_train)
auc_test_lr = roc_auc_score(y_cs_test, y_proba_test_lr)
print("Train AUC: %.4f" % auc_train)
print("Test AUC:  %.4f" % auc_test_lr)

# Convert predicted probabilities to binary (0 or 1) using threshold = 0.5
y_pred_train = (y_proba_train >= 0.13).astype(int)
y_pred_test = (y_proba_test_lr >= 0.13).astype(int)

# Calculate recall
recall_train = recall_score(y_cs_train, y_pred_train)
recall_test = recall_score(y_cs_test, y_pred_test)
print(f"Train Recall: {recall_train:.4f}")
print(f"Test Recall:  {recall_test:.4f}")

# Calculate precision
precision_train = precision_score(y_cs_train, y_pred_train)
precision_test = precision_score(y_cs_test, y_pred_test)
print(f"Train Precision: {precision_train:.4f}")
print(f"Test Precision:  {precision_test:.4f}")


In [ ]:
from sklearn.metrics import f1_score, precision_score, recall_score
import numpy as np

# Range of thresholds to test
thresholds = np.arange(0.0, 1.01, 0.01)

best_threshold = 0.5
best_f1 = 0

f1_scores = []

for thresh in thresholds:
    y_pred_test_thresh = (y_proba_test_lr >= thresh).astype(int)
    f1 = f1_score(y_cs_test, y_pred_test_thresh)
    f1_scores.append(f1)

    if f1 > best_f1:
        best_f1 = f1
        best_threshold = thresh

print(f" Best Threshold: {best_threshold:.2f}")
print(f" Best F1 Score: {best_f1:.4f}")

In [ ]:
# # Tuning Logistic Regression
# from sklearn.model_selection import GridSearchCV
# from sklearn.linear_model import LogisticRegression

# # Define the parameter grid
# param_grid = {
#     'C': [0.01, 0.1, 1, 10],              # Regularization strength (smaller = stronger)
#     'penalty': ['l1', 'l2'],              # Regularization type
#     'solver': ['liblinear'],              # 'liblinear' supports both l1 and l2
#     'max_iter': [1000]
#     }

# # Set up logistic regression with grid search and cross-validation
# log_reg = LogisticRegression()
# grid_search = GridSearchCV(log_reg, param_grid, cv=5, scoring='roc_auc', n_jobs=-1)

# # Fit on training data
# grid_search.fit(X_train, y_cs_train)

# # Best model
# best_model = grid_search.best_estimator_

# # Evaluate AUC
# y_proba_train = best_model.predict_proba(X_train)[:, 1]
# y_proba_test = best_model.predict_proba(X_test)[:, 1]

# auc_train = roc_auc_score(y_cs_train, y_proba_train)
# auc_test = roc_auc_score(y_cs_test, y_proba_test)

# print(" Best hyperparameters:", grid_search.best_params_)
# print("Train AUC: %.4f" % auc_train)
# print("Test AUC:  %.4f" % auc_test)


# Random Forest Classification

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.metrics import roc_auc_score
# Create Random Forest model
model = RandomForestClassifier(
    n_estimators=500, #300
    min_samples_leaf=1, #10
    max_features='sqrt',
    max_depth=10,
    random_state=42
)

# Fit model
model.fit(X_train, y_cs_train)

# Predict probabilities
y_proba_train = model.predict_proba(X_train)[:, 1]
y_proba_test_rf = model.predict_proba(X_test)[:, 1]

# AUC for train/test
auc_train = roc_auc_score(y_cs_train, y_proba_train)
auc_test_rf = roc_auc_score(y_cs_test, y_proba_test_rf)
print("Train AUC: %.4f" % auc_train)
print("Test AUC:  %.4f" % auc_test_rf)

# Convert predicted probabilities to binary (0 or 1) using threshold = 0.5
y_pred_train = (y_proba_train >= 0.2).astype(int)
y_pred_test = (y_proba_test_rf >= 0.2).astype(int)

# Calculate recall
recall_train = recall_score(y_cs_train, y_pred_train)
recall_test = recall_score(y_cs_test, y_pred_test)
print(f"Train Recall: {recall_train:.4f}")
print(f"Test Recall:  {recall_test:.4f}")

# Calculate precision
precision_train = precision_score(y_cs_train, y_pred_train)
precision_test = precision_score(y_cs_test, y_pred_test)
print(f"Train Precision: {precision_train:.4f}")
print(f"Test Precision:  {precision_test:.4f}")



In [ ]:
from sklearn.metrics import f1_score, precision_score, recall_score
import numpy as np

# Range of thresholds to test
thresholds = np.arange(0.0, 1.01, 0.01)

best_threshold = 0.5
best_f1 = 0

f1_scores = []

for thresh in thresholds:
    y_pred_test_thresh = (y_proba_test_rf >= thresh).astype(int)
    f1 = f1_score(y_cs_test, y_pred_test_thresh)
    f1_scores.append(f1)

    if f1 > best_f1:
        best_f1 = f1
        best_threshold = thresh

print(f"Best Threshold: {best_threshold:.2f}")
print(f"Best F1 Score: {best_f1:.4f}")

In [ ]:
# # Tuning Random Forest Classification
# from sklearn.model_selection import GridSearchCV
# from sklearn.ensemble import RandomForestClassifier

# param_grid = {
#     'n_estimators': [100, 200, 300],
#     'max_depth': [None, 10, 20],
#     'min_samples_leaf': [1, 5, 10],
#     'max_features': ['sqrt', 'log2']
# }

# grid_search = GridSearchCV(
#     estimator=RandomForestClassifier(random_state=42),
#     param_grid=param_grid,
#     scoring='roc_auc',
#     cv=5,
#     n_jobs=-1,
#     verbose=1
# )
# grid_search.fit(X_train, y_cs_train)

# print("Best params:", grid_search.best_params_)

# Neural Network

In [ ]:
from sklearn.model_selection import train_test_split, KFold
from sklearn.metrics import roc_auc_score
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam
import numpy as np
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.callbacks import EarlyStopping
import warnings
warnings.filterwarnings('ignore')


# Define model builder function
def create_model():
    model = Sequential([
        Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
        Dense(32, activation='relu'),
        Dense(1, activation='sigmoid')
    ])
    model.compile(optimizer=Adam(learning_rate=0.001),
                  loss='binary_crossentropy',
                  metrics=['AUC'])
    return model


# Early stop
early_stop = EarlyStopping(
    monitor='val_auc',
    patience=5,
    mode='max',
    restore_best_weights=True
)

# Create and fit model using validation split
model_vs = create_model()
model_vs.fit(X_train, y_cs_train, epochs=20, batch_size=64, validation_split=0.2, callbacks=[early_stop], verbose=0)

# Evaluate on held-out test set
y_proba_train = model_vs.predict(X_train).ravel()
y_proba_test_nn = model_vs.predict(X_test).ravel()
auc_train = roc_auc_score(y_cs_train, y_proba_train)
auc_test_nn = roc_auc_score(y_cs_test, y_proba_test_nn)
print("Validation Set Approch AUC — Train: %.4f" % auc_train)
print("Validation Set Apporch AUC — Test:  %.4f" % auc_test_nn)

# Convert predicted probabilities to binary (0 or 1) using threshold = 0.5
y_pred_train = (y_proba_train >= 0.15).astype(int)
y_pred_test = (y_proba_test_nn >= 0.15).astype(int)

# Calculate recall
recall_train = recall_score(y_cs_train, y_pred_train)
recall_test = recall_score(y_cs_test, y_pred_test)
print(f"Train Recall: {recall_train:.4f}")
print(f"Test Recall:  {recall_test:.4f}")

# Calculate precision
precision_train = precision_score(y_cs_train, y_pred_train)
precision_test = precision_score(y_cs_test, y_pred_test)
print(f"Train Precision: {precision_train:.4f}")
print(f"Test Precision:  {precision_test:.4f}")



In [ ]:
from sklearn.metrics import f1_score, precision_score, recall_score
import numpy as np

# Range of thresholds to test
thresholds = np.arange(0.0, 1.01, 0.01)

best_threshold = 0.5
best_f1 = 0

f1_scores = []

for thresh in thresholds:
    y_pred_test_thresh = (y_proba_test_nn >= thresh).astype(int)
    f1 = f1_score(y_cs_test, y_pred_test_thresh)
    f1_scores.append(f1)

    if f1 > best_f1:
        best_f1 = f1
        best_threshold = thresh

print(f"Best Threshold: {best_threshold:.2f}")
print(f"Best F1 Score: {best_f1:.4f}")

In [ ]:
# # Tuning Neural Network
# # -------------------------------
# # Define the Model Builder Function
# # -------------------------------
# from tensorflow.keras.layers import Dropout
# import itertools


# def build_model(units, dropout_rate, learning_rate, num_layers):
#     model = Sequential()
#     # Input layer.
#     model.add(Dense(units=units, activation='relu', input_shape=(X.shape[1],)))

#     # Hidden layers.
#     for _ in range(num_layers):
#         model.add(Dense(units=units, activation='relu'))
#         if dropout_rate > 0:
#             model.add(Dropout(rate=dropout_rate))

#     # Output layer for regression.
#     model.add(Dense(1, activation='sigmoid'))

#     # Compile the model using Adam optimizer and the custom Tweedie loss.
#     model.compile(optimizer=Adam(learning_rate=learning_rate),
#                   loss='binary_crossentropy',
#                   metrics=['AUC'])
#     return model

# # -------------------------------
# # Define Hyperparameter Grid
# # -------------------------------

# param_grid = {
#     'units':        [64, 128],
#     'dropout_rate': [0.0, 0.3],
#     'learning_rate': [0.001, 0.0005],  # ← decimals
#     'num_layers':   [2, 3, 4],
#     'epochs':       [20, 50],    # use EarlyStopping
#     'batch_size':   [32, 64]
# }

# # Create a list of all hyperparameter combinations.
# keys, values = zip(*param_grid.items())
# combinations = [dict(zip(keys, combination)) for combination in itertools.product(*values)]
# print("Total parameter combinations:", len(combinations))

# # -------------------------------
# # k-Fold Cross Validation
# # -------------------------------
# from sklearn.model_selection import KFold

# kf = KFold(n_splits=3, shuffle=True, random_state=42)

# best_score = np.inf  # Lower loss is better
# best_params = None

# # Iterate over all parameter combinations.
# for comb in combinations:
#     cv_losses = []
#     print("Evaluating combination:", comb)

#     # For each combination, perform cross-validation.
#     y_array = y_cs_train.to_numpy()
#     for train_index, val_index in kf.split(X_train):
#         X_tr, X_val = X_train[train_index], X_train[val_index]
#         y_tr, y_val = y_array[train_index], y_array[val_index]

#         # Build and train the model on the current training fold.
#         model = build_model(units=comb['units'],
#                             dropout_rate=comb['dropout_rate'],
#                             learning_rate=comb['learning_rate'],
#                             num_layers=comb['num_layers'])

#         model.fit(X_tr, y_tr,
#                   epochs=comb['epochs'],
#                   batch_size=comb['batch_size'],
#                   verbose=0)

#         # Evaluate on the validation fold.
#         loss = model.evaluate(X_val, y_val, verbose=0)
#         cv_losses.append(loss)

#     avg_cv_loss = np.mean(cv_losses)
#     print("Average CV Loss:", avg_cv_loss)

#     # Check if this combination has the best (lowest) loss so far.
#     if avg_cv_loss < best_score:
#         best_score = avg_cv_loss
#         best_params = comb.copy()

# print("\nBest CV Loss:", best_score)
# print("Best hyperparameters:", best_params)


#XGBoost

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.metrics import roc_auc_score, classification_report, recall_score, precision_score
import numpy as np
import xgboost as xgb

# Create the XGBoost model
model = xgb.XGBClassifier(
    objective='binary:logistic',
    eval_metric='logloss',
    max_depth=5, #3
    min_child_weight=5, #5
    subsample=0.8, #0.8
    colsample_bytree=1, #0.8
    learning_rate=0.05, #0.05
    n_estimators=200, #200
    random_state=42
)


# Fit model
model.fit(X_train, y_cs_train)

# Predict probabilities
y_proba_train = model.predict_proba(X_train)[:, 1]
y_proba_test_xgb = model.predict_proba(X_test)[:, 1]

# AUC for train/test
auc_train = roc_auc_score(y_cs_train, y_proba_train)
auc_test_xgb = roc_auc_score(y_cs_test, y_proba_test_xgb)
print("Train AUC: %.4f" % auc_train)
print("Test AUC:  %.4f" % auc_test_xgb)

# Convert predicted probabilities to binary (0 or 1) using threshold = 0.5
y_pred_train = (y_proba_train >= 0.15).astype(int)
y_pred_test = (y_proba_test_xgb >= 0.15).astype(int)

# Calculate recall
recall_train = recall_score(y_cs_train, y_pred_train)
recall_test = recall_score(y_cs_test, y_pred_test)
print(f"Train Recall: {recall_train:.4f}")
print(f"Test Recall:  {recall_test:.4f}")

# Calculate precision
precision_train = precision_score(y_cs_train, y_pred_train)
precision_test = precision_score(y_cs_test, y_pred_test)
print(f"Train Precision: {precision_train:.4f}")
print(f"Test Precision:  {precision_test:.4f}")



In [ ]:
from sklearn.metrics import f1_score, precision_score, recall_score
import numpy as np

# Range of thresholds to test
thresholds = np.arange(0.0, 1.01, 0.01)

best_threshold = 0.5
best_f1 = 0

f1_scores = []

for thresh in thresholds:
    y_pred_test_thresh = (y_proba_test_xgb >= thresh).astype(int)
    f1 = f1_score(y_cs_test, y_pred_test_thresh)
    f1_scores.append(f1)

    if f1 > best_f1:
        best_f1 = f1
        best_threshold = thresh

print(f"Best Threshold: {best_threshold:.2f}")
print(f"Best F1 Score: {best_f1:.4f}")

In [ ]:
# # Tuning XGBoost
# from sklearn.model_selection import GridSearchCV

# # Define hyperparameter grid
# param_grid = {
#     'max_depth': [3, 5],
#     'min_child_weight': [1, 5],
#     'subsample': [0.8, 1.0],
#     'colsample_bytree': [0.8, 1.0],
#     'learning_rate': [0.01, 0.05],
#     'n_estimators': [100, 200]
# }

# # Initialize base model
# xgb_clf = xgb.XGBClassifier(
#     objective='binary:logistic',
#     eval_metric='logloss',
#     random_state=42
# )

# # Set up GridSearchCV
# grid_search = GridSearchCV(
#     estimator=xgb_clf,
#     param_grid=param_grid,
#     scoring='roc_auc',
#     cv=5,
#     verbose=1,
#     n_jobs=-1
# )

# # Fit GridSearchCV
# grid_search.fit(X_train, y_cs_train)

# # Retrieve the best model
# best_model = grid_search.best_estimator_

# # Predict and evaluate
# y_proba_train = best_model.predict_proba(X_train)[:, 1]
# y_proba_test = best_model.predict_proba(X_test)[:, 1]

# auc_train = roc_auc_score(y_cs_train, y_proba_train)
# auc_test = roc_auc_score(y_cs_test, y_proba_test)

# print("Best Hyperparameters:", grid_search.best_params_)
# print("Train AUC: %.4f" % auc_train)
# print("Test AUC:  %.4f" % auc_test)


# LightGBM - Final

In [ ]:
from lightgbm import LGBMClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import roc_auc_score, recall_score, precision_score
import warnings
warnings.filterwarnings('ignore')

# Instantiate the LGBMClassifier
lgbm_model = LGBMClassifier(
    objective='binary',
    random_state=42,
    n_estimators=1000,
    learning_rate=0.01,
    num_leaves = 18,
    min_data_in_leaf = 100,
    max_depth = 5,
    class_weight='balanced',
    verbosity=-1
)

lgbm_model.fit(X_train, y_cs_train)


# Predict probabilities for train and test sets
y_proba_train = lgbm_model.predict_proba(X_train)[:, 1]
y_proba_test_lgb = lgbm_model.predict_proba(X_test)[:, 1]

# Compute ROC AUC
train_auc = roc_auc_score(y_cs_train, y_proba_train)
test_auc_lgb = roc_auc_score(y_cs_test, y_proba_test_lgb)

# Display results
print(f"Train AUC: {train_auc:.4f}")
print(f"Test AUC:  {test_auc_lgb:.4f}")

# Convert predicted probabilities to binary (0 or 1) using threshold = 0.5
y_pred_train = (y_proba_train >= 0.63).astype(int)
y_pred_test = (y_proba_test_lgb >= 0.63).astype(int)

# Calculate recall
recall_train = recall_score(y_cs_train, y_pred_train)
recall_test = recall_score(y_cs_test, y_pred_test)
print(f"Train Recall: {recall_train:.4f}")
print(f"Test Recall:  {recall_test:.4f}")

# Calculate precision
precision_train = precision_score(y_cs_train, y_pred_train)
precision_test = precision_score(y_cs_test, y_pred_test)
print(f"Train Precision: {precision_train:.4f}")
print(f"Test Precision:  {precision_test:.4f}")


In [ ]:
from sklearn.metrics import f1_score, precision_score, recall_score
import numpy as np

# Range of thresholds to test
thresholds = np.arange(0.0, 1.01, 0.01)

best_threshold = 0.5
best_f1 = 0

f1_scores = []

for thresh in thresholds:
    y_pred_test_thresh = (y_proba_test_lgb >= thresh).astype(int)
    f1 = f1_score(y_cs_test, y_pred_test_thresh)
    f1_scores.append(f1)

    if f1 > best_f1:
        best_f1 = f1
        best_threshold = thresh

print(f" Best Threshold: {best_threshold:.2f}")
print(f" Best F1 Score: {best_f1:.4f}")

In [ ]:
# # model tuning
# from lightgbm import LGBMClassifier
# from sklearn.model_selection import GridSearchCV
# from sklearn.metrics import roc_auc_score

# # Define the hyperparameter grid
# lgbm_params = {
#     'num_leaves': [18, 63],
#     'min_child_samples': [50, 100],
#     'max_depth': [5, 10],
#     'learning_rate': [0.01, 0.05],
# }

# # Instantiate the LGBMClassifier
# lgbm_model = LGBMClassifier(
#     objective='binary',
#     random_state=42,
#     n_estimators=1000
# )

# # Set up GridSearchCV
# grid_cv = GridSearchCV(
#     estimator=lgbm_model,
#     param_grid=lgbm_params,
#     scoring='roc_auc',
#     cv=5,
#     verbose=1,
#     n_jobs=-1
# )

# # Fit the model
# grid_cv.fit(X_train, y_cs_train)

# # Print best parameters
# print("Best hyperparameters:", grid_cv.best_params_)

# # Predict probabilities for train and test sets
# y_proba_train = grid_cv.predict_proba(X_train)[:, 1]
# y_proba_test = grid_cv.predict_proba(X_test)[:, 1]

# # Compute ROC AUC
# train_auc = roc_auc_score(y_cs_train, y_proba_train)
# test_auc = roc_auc_score(y_cs_test, y_proba_test)

# # Display results
# print(f"Train AUC: {train_auc:.4f}")
# print(f"Test AUC:  {test_auc:.4f}")


# Variable selection with SHAP


In [ ]:
# # Define X and Y then split and standardize###
# X = df_train[['X.9', 'X.10', 'X.11', 'X.12', 'X.14', 'X.22', 'X.23', 'X.24', 'X.25',
#        'X.26', 'X.28', 'age', 'driving_experience', 'policy_duration_days',
#        'time_since_last_renewal_days', 'X.7_1', 'X.13_1', 'X.19_2', 'X.19_3',
#        'X.19_4', 'X.20_1', 'X.21_1', 'X.27_P']]

# # Split into train and test set
# X_train_cs, X_test_cs = train_test_split(X, test_size=0.2, random_state=42)

# # y train cols
# y_cs_train = df_train.loc[X_train_cs.index,'CS']

# # y test cols
# y_cs_test = df_train.loc[X_test_cs.index,'CS']

# # Standardized X
# scaler = StandardScaler()
# X_train = scaler.fit_transform(X_train_cs)
# X_test = scaler.fit_transform(X_test_cs)

In [ ]:
#SHAP for CS with LightGBM
import shap

# Initialize SHAP explainer for LightGBM
explainer_cs = shap.Explainer(lgbm_model)

# Compute SHAP values
shap_values_cs = explainer_cs(X_train)

# Plot summary
shap.summary_plot(shap_values_cs, X_train, plot_type="bar")

# Detailed beeswarm SHAP plot
shap.summary_plot(shap_values_cs, X_train)

# ROC-AUC

In [ ]:
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, roc_auc_score

# Dictionary to store model names and predicted probabilities
models = {
    "Logistic Regression": y_proba_test_lr,
    "Random Forest": y_proba_test_rf,
    "XGBoost": y_proba_test_xgb,
    "LightGBM": y_proba_test_lgb,
    'Neural Network': y_proba_test_nn
}

plt.figure(figsize=(8, 6))

# Plot ROC curves
for name, y_proba in models.items():
    fpr, tpr, _ = roc_curve(y_cs_test, y_proba)
    auc = roc_auc_score(y_cs_test, y_proba)
    plt.plot(fpr, tpr, label=f"{name} (AUC = {auc:.2f})")

# Plot random baseline
plt.plot([0, 1], [0, 1], 'k--', label="Random")

# Formatting
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate (Recall)")
plt.title("ROC Curve Comparison")
plt.legend(loc="lower right")
plt.grid(True)
plt.tight_layout()
plt.show()


# Final Answer

In [ ]:
# Convert ['X.2', 'X.3', 'X.4', 'X.5', 'X.6'] in to number
# One-hot encoding ['X.7', 'X.13', 'X.19', 'X.20', 'X.21', 'X.27']

# Convert date columns to datetime
date_cols = ['X.2', 'X.3', 'X.4', 'X.5', 'X.6']
for col in date_cols:
    df_test[col] = pd.to_datetime(df_test[col], errors='coerce', dayfirst=True)

# Feature engineering
df_test['vehicle_age'] = 2020 - df_test['X.22']                    # from registration year
df_test['age'] = 2020 - df_test['X.5'].dt.year                     # from birthdate
df_test['driving_experience'] = 2020 - df_test['X.6'].dt.year     # from license issue date
df_test['policy_duration_days'] = (df_test['X.4'] - df_test['X.2']).dt.days
df_test['time_since_last_renewal_days'] = (df_test['X.3'] - df_test['X.2']).dt.days

# Drop raw date columns 'X.2', 'X.3', 'X.4', 'X.5', 'X.6' from test
df_test.drop(columns=[col for col in date_cols if col in df_test.columns], inplace=True)

# One-hot encoding
categorical_cols = ['X.7', 'X.13', 'X.19', 'X.20', 'X.21', 'X.27']
df_test = pd.get_dummies(df_test, columns=categorical_cols, drop_first=True)

In [ ]:
# Set df_test as feature then standardized
X = df_test
X_scale = scaler.fit_transform(X)

In [ ]:
df=df_train
# Drop raw dates and leakage columns
drop_cols = ["LC", "HALC", "CS"]

# Filter out extreme outliers
q_lc = df["LC"].quantile([0.001, 0.999])
q_halc = df["HALC"].quantile([0.001, 0.999])
#q_lc = df["LC"].quantile([0.01, 0.99])
#q_halc = df["HALC"].quantile([0.01, 0.99])
df_filtered = df[(df["LC"] >= q_lc.iloc[0]) & (df["LC"] <= q_lc.iloc[1]) &
               (df["HALC"] >= q_halc.iloc[0]) & (df["HALC"] <= q_halc.iloc[1])]

# Prepare features and targets
X_task1 = df_filtered.drop(columns=drop_cols)
#X = pd.get_dummies(X, columns=["X.27"], drop_first=True)
y_lc = df_filtered["LC"]
y_halc = df_filtered["HALC"]

In [ ]:
import pandas as pd
import lightgbm as lgb

# Step 3: Predict LC and HALC using trained LightGBM models

# Assume X is your training data (features only), y_lc and y_halc are the targets
# Refit final models on the full training data
params = {
    'n_estimators': 100,
    'max_depth': 7,
    'learning_rate': 0.05,
    'random_state': 42
}

# Train final models
model_lc = lgb.LGBMRegressor(**params)
model_lc.fit(X_task1, y_lc)

model_halc = lgb.LGBMRegressor(**params)
model_halc.fit(X_task1, y_halc)

# Predict
lc_pred = model_lc.predict(df_test)
halc_pred = model_halc.predict(df_test)

lc_pred = np.clip(lc_pred, 0, None)
halc_pred = np.clip(halc_pred, 0, None)

In [ ]:
# Fit X in to final model for CS
y_proba_test_lgb = lgbm_model.predict_proba(X_scale)[:, 1]
y_proba_test_lgb

In [ ]:
# Save all answer as df
final_result = pd.DataFrame({
                             'LC': lc_pred,
                             'HALC': halc_pred,
                             'CS': y_proba_test_lgb
                           })
final_result

In [ ]:
# # download file to local
# final_result.to_csv('group_21_prediction.csv', index=False)
# from google.colab import files
# files.download('group_21_prediction.csv')

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import pandas as pd

# Change path-to-file
df = pd.read_csv('/content/group_21_prediction.csv')

print(f"Number of rows:&nbsp;{len(df)}")
print(f"Number of columns:&nbsp;{len(df.columns)}")
print(f"Column names:&nbsp;{list(df.columns)}")
print(df.dtypes)